### DISCUSSION TITLE
Question about MOOSE Creep

### DISCUSSION THREAD

**Discussion by echo1115** [idaholab/moose#18254] [Q&A Modules: Solid mechanics] [answered]
Hi all:
    Based on RadialReturnCreepStressUpdateBase  or  PowerLawCreepStressUpdate in MOOSE, we can realize creep calculation related to Von Mises stress . I want to know how to  realize  the creep calculation in MOOSE if **the creep rate** is related to **the principal stress** but not to Von Mises stress.

Thanks a lot!

**Comment by ngrilli** [comment, marked as answer]
Dear @echo1115 

I think you need to change the expression for "effective_trial_stress" in RadialReturnStressUpdate
You can create another class with a different expression for the trial stress,
then two new classes derived from it like RadialReturnCreepStressUpdateBase and  PowerLawCreepStressUpdate

Nicolò

**Reply by echo1115** [reply]
Dear @ngrilli 

  Thank you for your reply. 
  Actually the creep rate I am considering is related to  the principal stresses in three directions, i.e. **έ=f(σ1,σ2,σ3)**.  
The question may not be solved by just changing the expression for "effective_trial_stress" in RadialReturnStressUpdate.
  Besides, i tried to realize this creep calculation by creating a  new class **"CreepTest"** derived from **RadialReturnCreepStressUpdateBase**  as shown in the following codes，but the calculation didn't converge.
```
Real
CreepTest::computeResidual(const Real effective_trial_stress, const Real scalar)
{
  _stress1[_qp] = RankTwoScalarTools::maxPrincipal(_stress[_qp], _input_direction);
  _stress2[_qp] = RankTwoScalarTools::midPrincipal(_stress[_qp], _input_direction);
  _stress3[_qp] = RankTwoScalarTools::minPrincipal(_stress[_qp], _input_direction);

  _creep_rate[_qp] = function(_stress1[_qp] ,_stress2[_qp] ,_stress3[_qp] )
  return _creep_rate[_qp] - scalar/_dt;
}

Real
CreepTest::computeDerivative(const Real effective_trial_stress, const Real scalar)
{
  return 0.0 - 1/_dt;
}
```

**Reply by hugary1995** [reply]
Right, I was expecting this. The isotropic creep model in tensor_mechanics assumes J2 plasticity. So if you modify the expression for effective trial stress, you also need to modify the discrete update. Your expression is equivalent to having a flow direction as a function of the spectrum of the stress, from which the discrete update should follow. There might be something tricky to implement though, because your flow direction isn't constant during the plasticity update.

**Reply by jiangwen84** [reply]
I am not sure exactly which creep model are you using. For some cases, e.g. eqn 15 in https://www.sciencedirect.com/science/article/pii/S0022311521000180 it is possible to derive an analytical expression in principal coordinate system, and then transform it back to the global coordinate system. For general case, you need to use a return mapping algorithm as suggested by @hugary1995  and @ngrilli

**Reply by echo1115** [reply]
> I am not sure exactly which creep model are you using. For some cases, e.g. eqn 15 in https://www.sciencedirect.com/science/article/pii/S0022311521000180 it is possible to derive an analytical expression in principal coordinate system, and then transform it back to the global coordinate system. For general case, you need to use a return mapping algorithm as suggested by @hugary1995 and @ngrilli

Dear @jiangwen84 and @hugary1995:
Thank you for all your reply.
According to the suggestion of @jiangwen84 in https://github.com/idaholab/moose/discussions/18254#discussioncomment-970813 , i read the paper by Wen Jiang et.al.  in https://www.sciencedirect.com/science/article/pii/S0022311521000180 and saw the **Eq.(15)**,
 i.e. _έ<sub>creep</sub>=K*( σ<sub>1</sub>-0.5(σ<sub>2</sub>+σ<sub>3</sub>) )*F_. where K and F are parameters depending on temperature and neutron fluence, σ<sub>i</sub> are the principal stresses.

If we want to realize the creep calculation in  **Eq.(15)**, i am not sure whether the "analytical expression in principal coordinate system" mentioned by @jiangwen84 in https://github.com/idaholab/moose/discussions/18254#discussioncomment-970813 refers to  **Eq.(15)** and then we should transform it  to the global coordinate system.
Besides, whether the newly created creep class still inherits from **RadialReturnCreepStressUpdateBase** or **ComputeEigenstrainBase** in MOOSE?

**Reply by ghost** [reply]
Hello @echo1115 did you find a solution to this problem?
